### PR TITLE
[WIP]Exclude resource alias for child

### PR DESCRIPTION
**Pull request by OptimusCrime**
### What does it do?
Implement functionality to hide a alias in the path of children resources.

Includes:

- This PR makes sure to add the new field in the map files of both MSSQL and MySQL. I've also added scripts for each of them to be added during an upgrade. This is test and works as expected.
- Adds a new checkbox under Settings in the box at the right side, below Searchable and above Freeze URI.
- Adds a simple lexicon entry with a wording similar to that in Evo. Appears when hoving the checkbox like the other settings in this area.

It works like this. Say we have the current relationship of resources Foo -> Bar, which has aliases `foo` and Bar which has its alias to `bar`. The default value of `alias_visible` is 1, so it does not change the current behavior of any sites without alterations. If both Foo and Bar has this setting to 1 (as is the default), it would produce the URI as before `foo/` and `foo/bar`. Below are the other configurations and their URI values:

Foo `alias_visible` = 0. Bar `alias_visible` = 1
Foo URI = `foo/`. Bar URI = `bar`.

Foo `alias_visible` = 1. Bar `alias_visible` = 0
Foo URI = `foo/`. Bar URI = `foo/bar`.

Foo `alias_visible` = 0. Bar `alias_visible` = 0
Foo URI = `foo/`. Bar URI = `bar`.

As we see, the URIs ignore the PARENT alias if `alias_visible` is 0, but will always use its own alias regardless of the `alias_visible` value. This is because we always require an alias for any resource.

*Note: This PR is currently missing tests. I'd love to implement some tests to make sure this works like it should.*

### Why is it needed?
Could be useful in certain situations. See issue referenced below. Seem like a common request. The feature was present in Evo, but for some reason removed since then. I see no reason this should not be included in the core. The functionality to go around this via an extra/plugin would be rather painful to implement.

### Related issue(s)/PR(s)
#11153

### CODE REVIEW

**Comment by Pathologic** [comment]
Thank you, I hope this try will be finished successfully.

**Comment by joeke** [comment]
@OptimusCrime 
Looking good, but I do have 2 comments/questions:
1. I think the field also has to be added to the modx.mysql.schema.xml (and sqlsrv.schema)
2. Does this work in combination with the `use_frozen_parent_uris` system setting? So let's say I have a 3 levels of resources:
+ Foo (alias_visible = 1)
++ Bar (alias_visible = 0, uri_override = 1)
+++Baz (alias_visible = 1)
In this case how will the uri of Baz look like, assuming you have `use_frozen_parent_uris` set to Yes?

**Comment by OptimusCrime** [comment]
@joeke Thank you for your input :) I'll add the fields to the schemas. 

If you look at the code for the implementation, it is [rather simple](https://github.com/OptimusCrime/revolution/blob/243e4ae5f929b65809f9f6490d399648cb86f00e/core/model/modx/modresource.class.php#L923-L927). No matter what the alias for a parent resource is, or how it was computed, it will not be added to the path if `alias_visible` is 0. 

So, based on this I'd guess the calculated URI for Baz would be `/foo/baz`. Do you think it should be different?

**Comment by pixelchutes** [comment]
I think I have the same question as @joeke -- how are conflicts handled when a parent container's alias is omitted from the path, and you are facing a frozen or pre-existing URI conflict? 

From your 3rd example above (both `alias_visible` = 0), what if you already had a Resource with `alias`=`bar` one level up, resulting in same URI? Likewise, what if some other Resource (regardless of location in hierarchy) had activated `uri_override` resulting in similar conflict?

Not opposed to the idea or progress at all, just trying to better understand!

**Comment by OptimusCrime** [comment]
This PR only alters the generation of the alias path. This means that other checks, such as duplicates, will function just as they did previously. I have no tested this, but it *should*.

Therefore @pixelchutes , both your questions would, as far as I can understand, result in duplicated URIs. This is simply because the values for the associated resource in the `site_content` table, after they are constructed, will be the same.

I am not sure if this is unwanted or not. The idea of this is really just to provide (yet) another way of handing aliases to best suit your needs. It will "conflict" with existing solutions such as `uri_override` and `use_forzen_parent_uris`, but as far as I can understand, those two options can already conflict among themselves.

Any input on alternative approaches to this is much appreciated. I did not browse the Evolution source code to see how this feature was implemented there, so I dont know if they have some of the same issues there.

**Comment by pixelchutes** [comment]
If that's how it will work, that's awesome! My line of thinking was if an _established_ site hierarchy "flipped the switch" at a certain level, you could have a potential for conflicts from a totally different corner of your site. At that point, I'm not sure to what degree the potential `uri` conflict check handles, but at a minimum a Refresh URI and/or Clear Cache would report any dupes in the console like always.

**Comment by OptimusCrime** [comment]
Great :)

As far as I can tell, by looking at the method(s) that handles aliases, the Resource create/update form should highlight and report a duplicate URI for this resource, just like any other duplicated URI. That part of the code has not been altered, and those methods does not care how the aliases were produced, only if they are duplicated of each other or not.

**Comment by joeke** [comment]
So as far as I can see it won't add any new conflicts with existing methods.
But I do think a helpful addition is that the error message for duplicate uri's should be optimised. Say for instance you have parent Foo with child Bar, where Foo has alias_visible = 0, and you also have Bar in the root level, then saving this one has conflicting uri with Foo > Bar, but it's not directly apparent that it has to do with the alias_visible = 0 setting on Foo. So a clear help/error message would be very helpful in this case.

**Comment by Mark-H** [comment]
Is this still WIP or ready for review?

**Comment by OptimusCrime** [comment]
I'd still like to add some tests, but the functionality is OK. There's no need to do anything else with the code. Could it be merged without the tests and I can add them to my other PR #13231 ?

**Comment by opengeek** [comment]
This is breaking existing tests. If those can be fixed we could review this for merge.

**Comment by OptimusCrime** [comment]
@opengeek The tests failed because the version of PhpUnit we used were incompatible with PHP7. It was fixed recently. I will rebase and push. Just a FYI on why the tests are red.

**Comment by OptimusCrime** [comment]
@opengeek Rebased and pushed. Tests should be green now.

File: setup/includes/upgrades/mysql/2.6.0-pl.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+/*1
```
Comment (Alroniks) [inline setup/includes/upgrades/mysql/2.6.0-pl.php]: `1` here is obsolete

**Comment by OptimusCrime** [comment]
Closing this as #13660 superseed my PR with additional fixes and follow up that I was unable to do.

**Pull request closed without merging**